src/Formats/JSONExtractTree.cpp

```diff
@@ -1482,12 +1554,232 @@ class DynamicNode : public JSONExtractTreeNode<JSONParser>
-                return std::make_shared<DataTypeMap>(std::make_shared<DataTypeString>(), makeNullable(std::make_shared<DataTypeString>()));
+            case ElementType::OBJECT:
+            {
+                return std::make_shared<DataTypeObject>(DataTypeObject::SchemaFormat::JSON, max_dynamic_paths_for_object, max_dynamic_types_for_object);
+            }
+        }
+    }
+
+    size_t max_dynamic_paths_for_object;
+    size_t max_dynamic_types_for_object;
+
+    /// Avoid building JSONExtractTreeNode for the same data types on each row by using cache.
+    mutable std::unordered_map<String, std::unique_ptr<JSONExtractTreeNode<JSONParser>>> json_extract_nodes_cache;
+};
+
+template <typename JSONParser>
+class ObjectJSONNode : public JSONExtractTreeNode<JSONParser>
+{
+public:
+    ObjectJSONNode(
+        std::unordered_map<String, std::unique_ptr<JSONExtractTreeNode<JSONParser>>> typed_path_nodes_,
+        const std::unordered_set<String> & paths_to_skip_,
+        const std::vector<String> & path_regexps_to_skip_,
+        size_t max_dynamic_paths_,
+        size_t max_dynamic_types)
+        : typed_path_nodes(std::move(typed_path_nodes_))
+        , paths_to_skip(paths_to_skip_)
+        , dynamic_node(std::make_unique<DynamicNode<JSONParser>>(
+              max_dynamic_paths_ / DataTypeObject::NESTED_OBJECT_MAX_DYNAMIC_PATHS_REDUCE_FACTOR,
+              std::max(max_dynamic_types / DataTypeObject::NESTED_OBJECT_MAX_DYNAMIC_TYPES_REDUCE_FACTOR, 1lu)))
+        , dynamic_serialization(std::make_shared<SerializationDynamic>())
+    {
+        sorted_paths_to_skip.assign(paths_to_skip.begin(), paths_to_skip.end());
+        std::sort(sorted_paths_to_skip.begin(), sorted_paths_to_skip.end());
+        for (const auto & regexp : path_regexps_to_skip_)
+            path_regexps_to_skip.emplace_back(regexp);
+    }
+
+    bool insertResultToColumn(IColumn & column, const typename JSONParser::Element & element, const JSONExtractInsertSettings & insert_settings, const FormatSettings & format_settings, String & error) const override
+    {
+        if (element.isNull() && format_settings.null_as_default)
+        {
+            column.insertDefault();
+            return true;
+        }
+
+        if (!element.isObject())
+        {
+            error = fmt::format("Cannot read JSON object from JSON element: {}", jsonElementToString<JSONParser>(element, format_settings));
+            return false;
+        }
+
+        auto & column_object = assert_cast<ColumnObject &>(column);
+        size_t prev_size = column_object.size();
+
+        /// Paths in shared data should be sorted, so we cannot insert paths there during traverse.
+        /// Instead we collect all paths and values that should go to shared data, sort them and insert later.
+        /// It's not optimal, but it's a price we pay for faster reading of subcolumns.
+        std::vector<std::pair<String, String>> paths_and_values_for_shared_data;
```

Comment:
I think both should have the same complexity. But in such cases I prefer vector because all the data is located in the same place in memory and iterating over it should be more optiman than iterating over data in map (as it's a tree where each node is allocated separately afaik).